Unit Tests + BugFixes

dino-rodriguez opened the pull request:
Thanks for the big initial push on the plugin @kincaidoneil!

- jest unit tests for all functions

- fixes for asym-server
  - fixed proxies not persisting account changes to store
  - fixed proxies not able to access plugin store
  - fixed fee estimation logic returns in gwei as expected (was previously wei)
  - fixed Trace class to properly catch when using 'throw Trace(..)'
  - fixed fulfill/reject timer on sending prepares
  - fixed formatAmount error on negative numbers
  - fixed channel state to be 0 not 1 for open channel
  - fixed wei <> gwei conversion missing in balance logic
  - fixed timer interval to be in milliseconds (previously seconds)
  - other minor stuff

sharafian commented:
This plugin is deprecated so I'm closing this. We can reopen if we ever need it in the future.

Commit:
Begin eth asym server tests

Write index, file, create simple instance, create simple memory store for
tests.

Commit:
test skeleton

Commit:
currency scale check

Commit:
create plugin base done

Commit:
Default framework and packages installed

Commit:
test: unit tests

Add jest unit tests for main plugin. WIP -- still have not finished.
Create simple memstore for tests. Jest config in package.json to work
properly with node (had a bug when not explicitly specifying node
environment).

Commit:
fix: various fixes

Fee estimation: now returns in gwei instead of wei as intended.

Account store: assign anonymous function to proxy to fix `this`
context (was not setting to plugin store). Also, set the object property
value on the set trap (was not persisting account changes to store).

Still a WIP, various TODOs throughout.

Commit:
test: more unit tests

Commit:
fix: trace class + timer race functioning

Throwing traces fixed (previously not catching). Replace ternary
operator logic of _handlePrepareResponse which was not working before.
Promise race in _handleCustomData was outputting invalid structure for
ILP reject, now properly returns.

sharafian commented on src/index.ts:
```diff
@@ -147,7 +152,8 @@ export default class MachinomyServerPlugin extends MiniAccountsPlugin {
       }, {
-        set (obj, prop, val) {
+        set: (obj, prop, val) => {
+          obj[prop] = val
```
This is redundant with the `Reflect.set` call. I'm also not familiar enough with Proxy/Reflect to know why this doesn't cause an infinite loop when we set inside of the set handler, but if the tests work then it must not be an issue

dino-rodriguez replied:
Will change to return true instead of the Reflect.set call -- forgot to update that. Previously, the object property was not set prior to putting the object in the store (thus storing an old object).

I'm not too familiar with proxies either, so I'm not sure why it doesn't cause an infinite loop. In the MDN example linked below, Reflect.set is used within the set handler of the proxy. Yet the interceptions section beneath the example states that Reflect.set triggers the trap as well.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/handler/set

sharafian commented on src/index.ts:
```diff
@@ -191,7 +197,7 @@ export default class MachinomyServerPlugin extends MiniAccountsPlugin {
       throw new Error(`Cannot connect to blocked account ${account.accountName}`)
     }
-
+ 
```
Looks like there's trailing whitespace here, is lint configured for this repo?

dino-rodriguez replied:
Handled, thanks. It is -- didn't run the linter before pushing, will do so next time.

sharafian commented on src/index.ts:
```diff
@@ -266,20 +271,21 @@ export default class MachinomyServerPlugin extends MiniAccountsPlugin {
       this._log.trace(`account ${account.accountName} debited ${formatAmount(amount, 'eth')}, new balance ${formatAmount(newBalance, 'eth')}`)
 
+      // Expiry timer for reject
+      let timer = new Promise(resolve => setTimeout(() => {
```
make sure to clear this timer after the Promise.race

sharafian commented on src/index.ts:
```diff
@@ -348,7 +354,7 @@ export default class MachinomyServerPlugin extends MiniAccountsPlugin {
         // isFunding should be toggled after all non-funding related aysnc operations have completed, but before any balance changes occur
         if (account.isFunding) {
-          throw new Trace(`Failed to pay account ${account.accountName}: another funding event was occuring simultaneously.`)
+          throw new Trace(`another funding event was occuring simultaneously.`)
```
I don't think an in-flight funding transaction should stop packets from being processed. Unless funding has a different meaning here than in the XRP context (adding funds to the channel) it should happen concurrently with traffic on the plugin. An in flight fund should just prevent more fund transactions from being issued until it's done.

sharafian commented on test/plugin.test.js:
```diff
@@ -0,0 +1,464 @@
+const Store = require('./util/memStore')
+const BtpPacket = require('btp-packet')
+const IlpPacket = require('ilp-packet')
+const debug = require('debug')('ilp-plugin-eohereum-asym-server:test')
```
eohereum -> ethereum

sharafian commented on test/plugin.test.js:
```diff
@@ -0,0 +1,464 @@
+
+/* TODO Mock for web3, HDWallet, and Machinomy API. */
+async function createPlugin() {
+  /* Silence console log output from provider creation. */
```
Can you make an upstream PR to use debug instead of console.log? It's quite strange for a library to use console.log without an option to turn it off

dino-rodriguez replied:
Submitted.

sharafian commented on test/plugin.test.js:
```diff
@@ -0,0 +1,464 @@
+
+beforeEach(async () => {
+  try {
+    this.plugin = await createPlugin() 
```
I'm not 100% sure how this works; `this` shouldn't be bound to anything in this context because it's an arrow function. I guess it's using the `this` of the global scope? Either way, functions in the tests should be changed to `function` instead of arrow function syntax.

dino-rodriguez replied:
The reason I did this is because Jest does not share the 'this' context across beforeEach, afterEach, and the test itself (as mocha does). 

https://github.com/facebook/jest/issues/3553

To get the shared variable access, I either had to use arrow functions with 'this', or define  variables as globals (as indicated above). I chose to use 'this' because I thought it would be visually more clear which variables were acting as globals. Can change it to the latter if it you think it is still best to have each test function individually scoped.

sharafian replied:
Ah, I see. I guess I'm just not as familiar with jest. If this is the common way that it's done then that's fine

sharafian commented on test/plugin.test.js:
```diff
@@ -0,0 +1,464 @@
+          expect(fee.toNumber()).toBeLessThan(1200000) 
+        }))
+      } catch (e) {
+        throw e
```
`catch(e) { throw e }` seems like a no-op

sharafian commented on test/plugin.test.js:
```diff
@@ -0,0 +1,464 @@
+      this.mockAccount.isBlocked = true
+      try {
+        await this.plugin._connect(this.mockAccountAddress, {})
+        throw Error('Test Error: should trigger catch')
```
Instead of doing this, you can do `expect(this.plugin._connect(this.mockAccountAddress, {})).to.be.rejectedWith('...')`

sharafian commented on test/plugin.test.js:
```diff
@@ -0,0 +1,464 @@
+          jest.runOnlyPendingTimers()
+          await protocolDataResponse
+          expect(oldBalance).toEqual(this.mockAccount.balance - this.mockIlpPrepare.amount)
+        } catch (e) {}
```
Why are errors being swallowed? that means this test isn't testing anything

dino-rodriguez replied:
The intention here (which I did in a couple of other places) was that for longer functions (like handlePrepareResponse), I wanted to test the status of something like a specific like an account's balance. Since it changed so many times over the course of the function, I wanted to capture it during a specific phase -- thus I caught errors which stopped the function early to test the status of the account balance at that point. Do you think it's best to forgoe these tests? I know unit testing logic should not peep inside functions too deeply, but I found this useful in finding a few bugs.

Commit:
chore: eslint and npm scripts

Commit:
fix: assorted bug fixes

* formatAmount fixed to handle negative numbers
* channel state should be of 0 not 1 for open channel
* added wei <> gwei conversion missing in balance logic
* timer adjusted from seconds to milliseconds

Commit:
test: all functions unit tested

Commit:
chore: install & configure eslint

kincaidoneil committed:
Update CircleCI config to run tests

kincaidoneil commented:
Circle on Node 10 is failing when it tries to build sqlite3, a Machinomy dependency--but merging this anyways since we're in the process of removing that.
